Add release notes for 2.3 release (#2771)
grafana/mimir: docs/sources/release-notes/v2.3.md

---
title: "Grafana Mimir version 2.3 release notes"
menuTitle: "V2.3 release notes"
description: "Release notes for Grafana Mimir version 2.3"
weight: 300
---

# Grafana Mimir version 2.3 release notes

Grafana Labs is excited to announce version 2.3 of Grafana Mimir, the most scalable, most performant open source time series database in the world.

The highlights that follow include the top features, enhancements, and bugfixes in this release. If you are upgrading from Grafana Mimir 2.2, there is upgrade-related information as well.
For the complete list of changes, see the [Changelog](https://github.com/grafana/mimir/blob/main/CHANGELOG.md).

> ## Features and enhancements

Reviewer: I think the following are worth a note:
- * [FEATURE] Compactor: Adds the ability to delete partial blocks after a configurable delay. This option can be configured per tenant. #2285
- * [FEATURE] Querier: Added support for tenant federation to metric metadata endpoint. #2467

Reviewer 2: Marco, what is the value of the first feature? Why would I want to delete partial blocks? Saves me more disk space? Are these partial blocks just bad blocks that were hanging around previously and never got cleaned up?

Reviewer 2: I think the second one is nice! 
@56quarters @pracucci maybe we could also update
https://grafana.com/docs/mimir/latest/operators-guide/securing/authentication-and-authorization/
to specify which endpoints do and don't support tenant federation. 

Today we have:

```You can federate queries across multiple tenants by using true in -tenant-federation.enabled=true. When you specify tenant IDs, separate them with a pipe (|) character in the ‘X-Scope-OrgID’ header, as in the example X-Scope-OrgID: tenant-1|tenant-2|tenant-3.
```

Ideally we'd specify this on our HTTP API docs instead of that page I mentioned above, labeling the APIs that support federation in the same way we label which APIs require the use of authentication. But that perhaps is a higher effort ask.

Reviewer: > Are these partial blocks just bad blocks that were hanging around previously and never got cleaned up?

This one. Compactor will complain about them forever, unless you manually do something (Mimir correctness is not affected). With the new feature, the compactor will auto-cleanup them.

Reviewer: > @56quarters @pracucci maybe we could also update
https://grafana.com/docs/mimir/latest/operators-guide/securing/authentication-and-authorization/
to specify which endpoints do and don't support tenant federation.

Good point. Updating the auth doc page LGTM, but I think longer term we should have a dedicated page for the query federation.

Reviewer 2: @pracucci re: partial block cleanup - is there a reason that it is off by default? Just because its experimental still? This seems like something that any user would want to turn on right? Since it would serve to cleanup these partial blocks that are corrupted/useless/leftover from some aborted compaction?

Reviewer 3: > Why would you want to ever keep those partial blocks around if they're useless?

It was a new feature, so we didn't enable it by default before testing on our servers.

Partial blocks can be a result of compactor interrupted during upload, or unfinished backfill upload. If user is not experiencing any of this, they don't need the feature.

I think having a default value for example `24h` would be OK.

Reviewer: > I think having a default value for example 24h would be OK.

We're still rolling out at Grafana Labs. I would wait next release before enabling it by default, but agree we should.

- **Support for OpenTelemetry Metrics ingestion path**:

>   Grafana Mimir now has experimental support for ingesting metrics from the [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/) via two options: `prometheusremotewrite` and `otlphttp`. There is also experimental support for the OpenTelemetry Metrics ingestion path on `/otlp/v1/metrics`.

Reviewer: Remote write was already supported. In this release we're introducing support for `otlphttp`.

```suggestion
  In addition to the [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/) exporter `prometheusremotewrite`, in this release Grafana Mimir introduces an experimental support for ingesting metrics using the `otlphttp` exporter. The `otlphttp` API endpoint exposed by Mimir is at the path `/otlp/v1/metrics`.
```

>   Grafana Mimir now has experimental support for ingesting metrics from the [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/) via two options: `prometheusremotewrite` and `otlphttp`. There is also experimental support for the OpenTelemetry Metrics ingestion path on `/otlp/v1/metrics`.

Reviewer 2: ```suggestion
This release of Grafana Mimir introduces experimental support for ingesting metrics from the [OpenTelemetry Collector's](https://opentelemetry.io/docs/collector/) [`otlphttp` exporter](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/otlphttpexporter/README.md). This adds a second ingestion option for users of the OTel Collector; Mimir was already compatible with the [`prometheusremotewrite` exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusremotewriteexporter). For more information, please see [Configure OTel Collector]({{< relref "../operators-guide/configure/configure-otel-collector.md">}}).
```

Reviewer 2: Let's see if my pr gets accepted :) 
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13434

> - **Added support to split instant queries by time**:

Reviewer: ```suggestion
- **Added experimental support to split instant queries by time**:
```

> - **Added support to split instant queries by time**:

Reviewer 2: ```suggestion
- **Increased instant query performance**:
```

>   As an experimental feature, instant queries in Grafana Mimir now will get the same benefits of query sharding we see for range queries.When enabled splitting instant queries by time will contribute in some cases to an exponential speed up in instant query response time.

Reviewer 4: ```suggestion
  As an experimental feature, instant queries in Grafana Mimir now will get the same benefits of query sharding we see for range queries. When enabled, splitting instant queries by time will contribute in some cases to an exponential speed up in instant query response time.
```

>   As an experimental feature, instant queries in Grafana Mimir now will get the same benefits of query sharding we see for range queries. When enabled, splitting instant queries by time will contribute in some cases to an exponential speed up in instant query response time.

Reviewer: ```suggestion
  As an experimental feature, instant queries in Grafana Mimir now will get the same benefits of query sharding we see for range queries. When enabled, splitting instant queries by time will contribute in some cases to a speed up in instant query response time.
```

>   As an experimental feature, instant queries in Grafana Mimir now will get the same benefits of query sharding we see for range queries. When enabled, splitting instant queries by time will contribute in some cases to an exponential speed up in instant query response time.

Reviewer 2: ```suggestion
Grafana Mimir added experimental support for splitting instant queries by time. This allows instant queries to benefit from query sharding in the same way range queries do. When this feature is enabled, users should see faster response times for certain instant queries. 
```

Reviewer 2: @pracucci wdyt?

> - **Object storage can now be configured for all components**:

Reviewer 2: ```suggestion
- **Simpler object storage configuration**:
```

>   Object storage can now be configured for all Grafana Mimir components using the `common` YAML config option key (or `-common.storage.*` CLI flags), this enhancement reduces the total configuration overhead for running Mimir with object storage and centralizes your object storage configuration in one place reducing errors.

Reviewer: ```suggestion
  Object storage can now be configured for all Grafana Mimir components using the `common` YAML config option key (or `-common.storage.*` CLI flags). This enhancement reduces the total configuration overhead for running Mimir with object storage and centralizes your object storage configuration in one place reducing errors.
```

>   Object storage can now be configured for all Grafana Mimir components using the `common` YAML config option key (or `-common.storage.*` CLI flags), this enhancement reduces the total configuration overhead for running Mimir with object storage and centralizes your object storage configuration in one place reducing errors.

Reviewer: I would also mention:
- The common config applies to blocks, alertmanager and ruler storage config
- You can still specify the individual configs, which overrides whatever set in the common one
- A link to the doc of the common config

>   Object storage can now be configured for all Grafana Mimir components using the `common` YAML config option key (or `-common.storage.*` CLI flags), this enhancement reduces the total configuration overhead for running Mimir with object storage and centralizes your object storage configuration in one place reducing errors.

Reviewer 2: ```suggestion
Users can now configure block, alertmanager, and ruler storage all at once with the `common` YAML config option key (or `-common.storage.*` CLI flags). By centralizing your object storage configuration in one place, this enhancement makes configuration faster and less error prone. Users can still individually configure storage for each of these components if they desire. For more information, see the [common section of the configuration reference.]({{< relref "../operators-guide/configure/reference-configuration-parameters/#common">}}). 
```

Reviewer 2: @pracucci wdyt? @colega I actually think what @pracucci mentioned should also go in more permanent configuration docs than just the release notes (which a user might only check upon upgrade to 2.3). 

In other words, I think his point about "you can still configure things individually" and "individual configs will override the common config" should be called out somewhere. Is that already done and I just missed it? 
One place we could put it is here: ../operators-guide/configure/reference-configuration-parameters/#common.

> - **Add packaging for both debian based deb file and redhat based rpm**:

Reviewer 4: Maybe we can call out that this was a community contribution?

> - **Add packaging for both debian based deb file and redhat based rpm**:

Reviewer 2: ```suggestion
- **DEB and RPM packages for Mimir **:
```

>   With Grafana Mimir 2.3 we will be publishing a deb file and a rpm file of Grafana Mimir to make installation and running on debian or redhat based systems simpler. This was made possible by a community contribution from wilfriedroset.

Reviewer: When the release note will be read, the release assets will be already published.

```suggestion
  With Grafana Mimir 2.3 we're publishing a deb file and a rpm file of Grafana Mimir to make installation and running on debian or redhat based systems simpler. This was made possible by a community contribution from wilfriedroset.
```

>   With Grafana Mimir 2.3 we will be publishing a deb file and a rpm file of Grafana Mimir to make installation and running on debian or redhat based systems simpler. This was made possible by a community contribution from wilfriedroset.

Reviewer: I would link `wilfriedroset` to their GitHub profile.

>   With Grafana Mimir 2.3 we will be publishing a deb file and a rpm file of Grafana Mimir to make installation and running on debian or redhat based systems simpler. This was made possible by a community contribution from wilfriedroset.

Reviewer 2: ```suggestion
Starting with version 2.3, we're publishing deb and rpm files for Grafana Mimir, which will make installing and running it on Debian or RedHat-based linux systems much easier. Thank you to community contributor [wilfriedroset](https://github.com/wilfriedroset) for your work to implement this! 
```

> - **Added `backfill` command to mimirtool to upload Prometheus TSDB blocks:**

Reviewer 2: ```suggestion
- **Import historic data to Grafana Mimir**
```

>   The `mimirtool backfill` command will make the process of back-filling Prometheus TSDB blocks using API available in the compactor even easier which allows for a migration to Grafana Mimir while keeping your previous Prometheus TSDB data.

Reviewer 4: ```suggestion
  The `mimirtool backfill` command will make the process of back-filling Prometheus TSDB blocks, using an API available in the compactor, even easier which allows for a migration to Grafana Mimir while keeping your previous Prometheus TSDB data.
```

>   The `mimirtool backfill` command will make the process of back-filling Prometheus TSDB blocks, using an API available in the compactor, even easier which allows for a migration to Grafana Mimir while keeping your previous Prometheus TSDB data.

Reviewer 2: ```suggestion
Users can now backfill time series data from their existing Prometheus or Cortex installation into Mimir using `mimirtool`, making it possible to migrate to Grafana Mimir without losing your existing metrics data. This support is still considered experimental and does not work for data stored in Thanos yet. To learn more about this feature, please see [mimirtool backfill]({{< relref "../operators-guide/tools/mimirtool/#backfill">}}) and [Configure TSDB block upload]({{< relref "../operators-guide/configure/configure-tsdb-block-upload/">}})
```

## Upgrade considerations

> In Grafana Mimir 2.3 we have removed the previously deprecated configuration options:

Reviewer 2: ```suggestion
In Grafana Mimir 2.3 we have removed the following previously deprecated configuration options:
```

- `extend_writes` YAML key and `-distributor.extend-writes` CLI flag from the distributor config.

> - `active_series_custom_trackers` YAML key from the ingester config.

Reviewer 4: Maybe expand on this a bit since it's now in the runtimeconfig only?

> - `extend_writes` YAML key and `-distributor.extend-writes` CLI flag from the distributor config.

Reviewer 2: ```suggestion
- The `extend_writes` parameter in the distributor YAML configuration and `-distributor.extend-writes` CLI flag have been removed.
```

> - `active_series_custom_trackers` YAML key from the ingester config as it has been moved to the runtime config. See [#1188](https://github.com/grafana/mimir/pull/1188) for details.

Reviewer 2: ```suggestion
- The `active_series_custom_trackers` parameter has been removed from the YAML configuration. It had already been moved to the runtime configuration. See [#1188](https://github.com/grafana/mimir/pull/1188) for details.
```

> With Grafana Mimir 2.3 we have also updated the default value for `-distributor.ha-tracker.max-clusters` to `100` to provide DoS protection.

Reviewer 2: ```suggestion
With Grafana Mimir 2.3 we have also updated the default value for `-distributor.ha-tracker.max-clusters` to `100` to provide Denial-of-Service protection.
```

> ## Bug fixes

Reviewer 2: Would be nice if there was a bit of a sense of the impact of these bugs. Like what was the previous behavior and what's the new behavior. 

For example:
Memberlist KV store now tries to "fast-join" the cluster to avoid serving empty KV store. You see see fewer memberlist restarts upon startup. 

^^I just made that up of course.

Reviewer 2: Or:

Do not block the Ruler's "List Prometheus rules" API endpoint while syncing rules. This previously caused deadlocks when the ruler was communicating with object storage.

The highlights that follow include the top features, enhancements, and bugfixes in this release. If you are upgrading from Grafana Mimir 2.2, there is upgrade-related information as well.
For the complete list of changes, see the [Changelog](https://github.com/grafana/mimir/blob/main/CHANGELOG.md).

## Features and enhancements

- **Ingest metrics in OpenTelemetry format**:
  This release of Grafana Mimir introduces experimental support for ingesting metrics from the [OpenTelemetry Collector's](https://opentelemetry.io/docs/collector/) [`otlphttp` exporter](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/otlphttpexporter/README.md). This adds a second ingestion option for users of the OTel Collector; Mimir was already compatible with the [`prometheusremotewrite` exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusremotewriteexporter). For more information, please see [Configure OTel Collector]({{< relref "../operators-guide/configure/configure-otel-collector.md">}}).

- **Increased instant query performance**

>   As an experimental feature, instant queries in Grafana Mimir now will get the same benefits of query sharding we see for range queries. When enabled, splitting instant queries by time will contribute in some cases to a speed up in instant query response time.

Author: [Not yet](https://github.com/grafana/mimir/issues/2767 ), I can add the yaml/cli config option though.

>   As an experimental feature, instant queries in Grafana Mimir now will get the same benefits of query sharding we see for range queries. When enabled with `-query-frontend.split-instant-queries-by-interval`, splitting instant queries by time will contribute in some cases to a speed up in instant query response time.

Reviewer 2: ```suggestion
Grafana Mimir now supports splitting instant queries by time. This allows it to better parallelize execution of instant queries and therefore return results faster. At present, splitting is only supported for a subset of instant queries, which means not all instant queries will see a speedup. This feature is being released as experimental and is disabled by default. It can be enabled by setting `-query-frontend.split-instant-queries-by-interval`.
```

  - Memberlist now uses DNS service-discovery by default, which should decrease startup time for large Mimir clusters.

## Upgrade considerations

In Grafana Mimir 2.3 we have removed the following previously deprecated configuration options:

- The `extend_writes` parameter in the distributor YAML configuration and `-distributor.extend-writes` CLI flag have been removed.
- The `active_series_custom_trackers` parameter has been removed from the YAML configuration. It had already been moved to the runtime configuration. See [#1188](https://github.com/grafana/mimir/pull/1188) for details.

With Grafana Mimir 2.3 we have also updated the default value for `-distributor.ha-tracker.max-clusters` to `100` to provide Denial-of-Service protection. Previously `-distributor.ha-tracker.max-clusters` was unlimited by default which could allow a tenant with HA Dedupe enabled to overload the HA tracker with `__cluster__` label values that could cause the HA Dedupe database to fail.

## Bug fixes

> - [PR 2447](https://github.com/grafana/mimir/pull/2447): Fix incorrect mapping of http status codes 429 to 500 when request queue is full in the query-frontend. This corrects behavior in the query-frontend where a 429 "Too Many Outstanding Requests" from a querier would be returned to the user as a 500 system error when the error is retryable.

Reviewer 2: ```suggestion
- [PR 2447](https://github.com/grafana/mimir/pull/2447): Fix incorrect mapping of http status codes `429` to `500` when the request queue is full in the query-frontend. This corrects behavior in the query-frontend where a `429 "Too Many Outstanding Requests"` error (a retriable error) from a querier was incorrectly returned as a `500` system error (an unretriable error).
```

> - [PR 2505](https://github.com/grafana/mimir/pull/2505): Memberlist KV store now tries to "fast-join" the cluster to avoid serving empty KV store. This fix is to address the confusing "empty ring" error response and the error log message "ring doesn't exist in KV store yet" from services when there are other members in the ring when a service starts.

Reviewer 2: ```suggestion
- [PR 2505](https://github.com/grafana/mimir/pull/2505): The Memberlist key<>value (KV) store now tries to "fast-join" the cluster to avoid serving an empty KV store. This fix addresses the confusing "empty ring" error response and the error log message "ring doesn't exist in KV store yet" emitted by services when there are other members present in the ring when a service starts. Those using other key<>value store options (e.g., consul, etcd) are not impacted by this bug. 
```

Reviewer: >key<>value

I would call it "key-value" as in the doc.

> - [PR 2289](https://github.com/grafana/mimir/pull/2289): Do not block the Ruler's "List Prometheus rules" API endpoint while syncing rules. This will allow users to be able to list rules while syncing larger rule sets.

Reviewer 2: ```suggestion
- [PR 2289](https://github.com/grafana/mimir/pull/2289): The "List Prometheus rules" API endpoint of the Mimir Ruler component is no longer blocked while rules are being synced. This means users can now list rules while syncing larger rule sets.
```

- **Increased instant query performance**
  As an experimental feature, instant queries in Grafana Mimir now will get the same benefits of query sharding we see for range queries. When enabled, splitting instant queries by time will contribute in some cases to a speed up in instant query response time.

- **Tenant federation for metadata queries**
  Previously with tenant federation enabled, instant queries, range queries, and exemplars could be queried for multiple tenant by supplying multiple tenant IDs. With Grafana Mimir 2.3 we've added tenant federation support querying the `/api/v1/metadata` endpoint with multiple tenant IDs.

- **Simpler object storage configuration**:

>   Users can now configure block, alertmanager, and ruler storage all at once with the `common` YAML config option key (or `-common.storage.*` CLI flags). By centralizing your object storage configuration in one place, this enhancement makes configuration faster and less error prone. Users can still individually configure storage for each of these components if they desire. For more information, see the [common section of the configuration reference.]({{< relref "../operators-guide/configure/reference-configuration-parameters/#common">}}).

Reviewer 2: ```suggestion
  Users can now configure block, alertmanager, and ruler storage all at once with the `common` YAML config option key (or `-common.storage.*` CLI flags). By centralizing your object storage configuration in one place, this enhancement makes configuration faster and less error prone. Users can still individually configure storage for each of these components if they desire. For more information, see the [common section of the configuration reference.]({{< relref "../operators-guide/configure/about-configurations/#common-configurations">}}).
```

This link is more informative than the previous one.
